 modules/lua/mod_lua.c | 7 +------
 1 file changed, 1 insertion(+), 6 deletions(-)

diff --git a/modules/lua/mod_lua.c b/modules/lua/mod_lua.c
index 459e252..69f23d8 100644
--- a/modules/lua/mod_lua.c
+++ b/modules/lua/mod_lua.c
@@ -41,11 +41,6 @@
 #include <unistd.h>
 #endif
 
-/* getpid for Windows */
-#if APR_HAVE_PROCESS_H
-#include <process.h>
-#endif
-
 APR_IMPLEMENT_OPTIONAL_HOOK_RUN_ALL(ap_lua, AP_LUA, int, lua_open,
                                     (lua_State *L, apr_pool_t *p),
                                     (L, p), OK, DECLINED)
@@ -119,11 +114,11 @@ static void lua_open_callback(lua_State *L, apr_pool_t *p, void *ctx)
     ap_lua_load_apache2_lmodule(L);
     ap_lua_load_request_lmodule(L, p);
     ap_lua_load_config_lmodule(L);
+    ap_lua_run_lua_open(L, p);
 }
 
 static int lua_open_hook(lua_State *L, apr_pool_t *p)
 {
-    lua_open_callback(L, p, NULL);
     return OK;
 }
 
